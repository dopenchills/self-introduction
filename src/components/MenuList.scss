@use '../color';

.menu-list {
  // flexbox-parent
  display: flex;
  flex-direction: column;
  flex-wrap: nowrap;
  justify-content: flex-end;
  align-items: stretch;
  align-content: stretch;
  gap: 0;
}

.menu-option {
  display: block;
  
  input {
    display: none;
  }

  label {
    // text-align
    display: block;   
    text-align: end;

    // color
    background-color: color.$whitish-color-light-half;

    padding: 0.5em 0.5em 0.5em 2em;

    user-select: none;

    &:hover {
      background-color: color.$whitish-color-light-quarter;
    }
  }

  input:checked + label {
    background-color: color.$whitish-color-light;
  }
}
