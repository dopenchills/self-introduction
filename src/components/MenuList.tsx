import * as React from "react"
import { PreferenceContext } from "./contexts/Preference"
import { MenuContext } from "./contexts/Menu"
import './MenuList.scss'

const MenuOption = (
  props: {
    className?: string,
    name?: string,
    value: string,
    label?: string,
    checked?: boolean,
  }): JSX.Element => {
  const { isLightMode } = React.useContext(PreferenceContext)
  const { dispatch } = React.useContext(MenuContext)

  const name                  = props.name   || "menu"
  const label                 = props.label  || props.value
  const id                    = `${name}-${props.value}`


  const onMouseEnter = () => {
    dispatch({type: "hovered", payload: props.value})
  }
  const onMouseLeave = () => {
    dispatch({type: "hovered", payload: ""})
  }
  return (
    <div className={`menu-option ${props.className}`} onMouseEnter={onMouseEnter} onMouseLeave={onMouseLeave} >
      <input type="radio" name={name} value={props.value} id={id} checked={props.checked} readOnly={true} />
      <label htmlFor={id} className={`${isLightMode ? "" : "dark-mode"}`}>{label}</label>
    </div>
  )
}

const MenuList = (
  props: {
    className?: string,
    value?: string,
  }): JSX.Element => {
    const { dispatch: dispatchPreference } = React.useContext(PreferenceContext)
    const { selected, dispatch: dispatchMenu } = React.useContext(MenuContext)
  const menuOptionValues = [
    "home",
    "skills",
    "career",
    "portfolio",
  ]

  const onMenuListChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    dispatchMenu({type: "selected", payload: event.target.value})
  }

  const onClick = () => {
    dispatchPreference({type: "hideMenuPane", payload: undefined})
  }
  
  const menuOptions = menuOptionValues.map(value => {
    if (value === selected) {
      return <MenuOption key={value} className="flex-item" name="menu" value={value} label={value.toUpperCase()} checked={true} />
    }
    return <MenuOption key={value} className="flex-item" name="menu" value={value} label={value.toUpperCase()} checked={false} />
  })

  return (
    <div className={`menu-list ${props.className}`} onChange={onMenuListChange} onClick={onClick}>
      {menuOptions}
    </div>
  )
}

export default MenuList
