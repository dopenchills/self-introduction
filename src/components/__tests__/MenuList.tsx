import * as React from "react"
import renderer from "react-test-renderer"
import * as Gatsby from "gatsby"
import {cleanup, fireEvent, render} from '@testing-library/react'

import { MenuOption } from "../MenuList"

describe(`MenuOption`, () => {
  it(`has the same snapshot`, () => {
    const component = renderer.create(
      <MenuOption name="menu" value="portfolio" label={"portfolio".toUpperCase()} checked={true} />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
