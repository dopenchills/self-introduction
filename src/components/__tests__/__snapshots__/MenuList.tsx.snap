// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuOption has the same snapshot 1`] = `
<div
  className="menu-option undefined"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <input
    checked={true}
    id="menu-portfolio"
    name="menu"
    readOnly={true}
    type="radio"
    value="portfolio"
  />
  <label
    htmlFor="menu-portfolio"
  >
    PORTFOLIO
  </label>
</div>
`;
